control/controlclient/map.go

```diff
@@ -343,6 +352,73 @@ func undeltaPeers(mapRes *tailcfg.MapResponse, prev []*tailcfg.Node) {
 	mapRes.PeersRemoved = nil
 }
 
+// For extra defense-in-depth, when we're testing expired nodes we check
+// ControlTime against this 'epoch' (set to the approximate time that this code
+// was written) such that if control (or Headscale, etc.) sends a ControlTime
+// that's sufficiently far in the past, we can safely ignore it.
+var removeExpiredPeersEpoch = time.Unix(1673373066, 0)
+
+// If the offset between the current time and the time received from control is
+// larger than this, we store an offset in our mapSession to adjust future
+// clock timings.
+const minClockDelta = 1 * time.Minute
+
+// removeExpiredPeers updates mapRes.Peers to remove all peers that have
+// expired per the ControlTime field in the MapResponse.
+//
+// This is a defense-in-depth against something going wrong with control such
+// that we start seeing expired peers; we'll drop them here, if so.
+func (ms *mapSession) removeExpiredPeers(mapRes *tailcfg.MapResponse) {
+	now := clockNow()
+
+	// If we have a ControlTime field, update our delta.
+	if mapRes.ControlTime != nil && !mapRes.ControlTime.IsZero() {
+		delta := mapRes.ControlTime.Sub(now)
+		if delta.Abs() > minClockDelta {
+			log.Printf("[v1] netmap: removeExpiredPeers: setting clock delta to %v", delta)
+			ms.clockDelta = delta
+		}
+	}
+
+	// Adjust our current time by any saved delta to adjust for clock skew.
+	now = now.Add(ms.clockDelta)
+	if now.Before(removeExpiredPeersEpoch) {
+		log.Printf("netmap: removeExpiredPeers: [unexpected] delta-adjusted current time is before hardcoded epoch; skipping")
+		return
+	}
+
+	var numExpired uint64
+	for _, peer := range mapRes.Peers {
+		// Nodes that don't expire have KeyExpiry set to the zero time
+		if peer.KeyExpiry.IsZero() || peer.KeyExpiry.After(now) {
+			continue
+		}
+
+		log.Printf("[v1] netmap: removeExpiredPeers: clearing expired peer %v", peer.StableID)
+		numExpired++
+
+		// Actually mark the node as expired
+		peer.Expired = true
+
+		// Control clears the Endpoints and DERP fields of expired
```

Comment:
add another sentence explaining that this is insufficient, though, due to endpoints-over-DERP and DERP discovery via DRPO?

Reply:
Done